### ISSUE TITLE
Documentation on re-exports is ignored if the item is defined in the same crate

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#84619]
I tried this code:

```rust
mod inner_mod {
    /// docs from inner module
    pub fn g() {}
}

/// docs from outer module
pub use inner_mod::g;
```

I expected to see this happen: The documentation is added, like it is for cross-crate re-exports:
![image](https://user-images.githubusercontent.com/23638587/116268750-9a6aac00-a74b-11eb-9a39-49b69445bd59.png)

```rust
/// Some docs from outer crate
pub use inner::f;
```

Instead, this happened: The documentation is ignored:

![image](https://user-images.githubusercontent.com/23638587/116268785-a22a5080-a74b-11eb-8e4e-cce8bbde483f.png)

### Meta

`rustdoc --version`: rustdoc 1.53.0-nightly (392ba2ba1 2021-04-17)

_Originally posted by @jsgf in https://github.com/rust-lang/rust/issues/84597#issuecomment-827215642_

**Comment by Lonami** [comment]
Appending the documentation seems slightly strange, at least for the same crate. My expectation is that the re-export would have its own and, if you navigate to the re-exported name, you see the original documentation.

**Comment by jyn514** [comment]
@Lonami concatenating them is what rustdoc does for cross crate links, it should do the same thing for intra crate links. We shouldn't change it now.

**Comment by jyn514** [comment]
Well, here's why: https://github.com/rust-lang/rust/blob/6c2dd251bbff03c7a3092d43fb5b637eca0810e3/src/librustdoc/clean/inline.rs#L49-L51

but removing that block breaks rustdoc quite a lot:
```
failures:
    [rustdoc] rustdoc/foreigntype-reexport.rs
    [rustdoc] rustdoc/inline_local/please_inline.rs
    [rustdoc] rustdoc/intra-doc/field.rs
    [rustdoc] rustdoc/proc-macro.rs
    [rustdoc] rustdoc/reexport-stability-tags-deprecated-and-portability.rs
    [rustdoc] rustdoc/reexport-stability-tags-unstable-and-portability.rs
```

**Comment by Lonami** [comment]
I'm not familiar with rustdoc tests, but looks like we should just updated the expected results for the tests (assuming we do want to do this change, of course). As in, I don't think things are "breaking" as signaled by the tests with this change, but rather it just needs updating.

**Comment by jyn514** [comment]
@Lonami no, there are larger changes, it's not just documentation being added. I don't have the exact output on hand, it was a lot.

**Comment by lambinoo** [comment]
I think I'll try to tackle this one! Maybe this will also fix #89020 
@rustbot claim
